#322 Set config store to be included by default (#329)
boozallen/aissemble: docs/modules/ROOT/pages/configuration-store.adoc

= Universal Configuration Store

== Overview
The Universal Configuration Store is a tool that enables the various configurations for a project to be centrally defined and managed. 
It then provides a standardized way of accessing them, allowing the environment specific configurations to be dynamically provided to 
their respective resources within the project at runtime.

=== Setup
To setup the configuration store, one must first create properties files for storing the configurations. Each individual configuration is 
represented as a property key value pair within the file. These files must then be made accessible to the configuration store kubernetes 

> resource for ingestion. See the applicable sections below for more details on defining configurations as properties, creating environment 

Reviewer: A: Any time we say "See XYZ..." we should just link the section we're talking about.  In-window links for any internal navigation and new-window links for any external navigation.  I'd also suggest where we can working the link more organically into a sentence rather than outright saying "See XYZ". E.g. The first sentence could read _"...must first [define the configurations] through properties files."_ where _define the configurations_ links to the relevant section.

Author: Update hyperlinks to flow with sentence

specific property overrides, mounting the properties, and specifying the configuration store backend storage type.

> ==== Defining Configuration Properties

Reviewer: S: I think we have an opportunity to offload a lot of this explanation by linking the Krausening docs.  I'd also suggest that for now we don't bring in the concept of a "group" until it's absolutely necessary (i.e. during the Usage docs where there is an explicit parameter call "group" that has to be provided.)

Properties represent key values pairs for storing the project configurations. The key values pairs are stored as string types using 
the standard `.properties` file format. Each property then has an associated group name determined by the name of the file. Properties
are then accessed using a combination of their group name and property name. Accessing the properties is covered in more detail in the 
<<Usage>> section below.

Here is an example of a base properties file named `messaging.properties`:
[source,properties]
----
connector=smallrye-kafka
topic=baseTopic
----
This file defines two properties, `connector` and `topic`, each with the group name `messaging`.

===== Environment Specific Overrides

> In addition to the base properties file introduced above, one can optionally specify a secondary properties file that houses environment

Reviewer: S: Again I think we can largely lean on Krausening docs here and only call out the aiSSEMBLE-specific stuff.

Author: Looking at the [krausening related section for this](https://github.com/TechnologyBrewery/krausening/tree/dev/krausening/), while it does explain how the extension works, I think it introduces too many unrelated details. Specifically the stuff the env var `KRAUSENING_EXTENSIONS` and the code block for retrieving the properties.

Would rather just keep a straightforward example here that just focus on the properties files.

specific configurations. These configurations will then used to override their respective base configurations.

> Continuing with the example above, suppose one defines an additional properties file named `messaging.properties` like the following:

Author: In theory they would need to be the same name but in different directories for the environment ovverride to work. This is no longer applicable though as I removed the property file names from this section to make it less confusing.

[source,properties]
----
topic=ciTopic
auth-token=auth-secret
----

Then when these two properties files are reconciled within the store, it results in a configuration equivalent to the following:
[source,properties]
----
connector=smallrye-kafka
topic=ciTopic
auth-token=auth-secret
----

NOTE: The `topic` property is overwritten when the environment specific value `ciTopic`.

> ==== Mounting Properties

Reviewer: S: I think this would be more consumable if the over-arching section was "Deploying" and we frame everything in that context.  We could have a quick intro paragraph that talks about/links the Helm chart documentation, and what the key goals are of the deployment setup (making the configuration properties files available to the service and choosing the backing store)

Author: Refactored

In order for the store to read the configurations, the properties files will need to be placed into a https://kubernetes.io/docs/concepts/storage/persistent-volumes/#lifecycle-of-a-volume-and-claimpersistent[Persistent Volume,role=external,window=_blank] (PV)

> mounted to the configuration store kubernetes resource. It is assumed a PV will meet most projects needs and is the most widely accepted 

Reviewer: S: We've got a fine line to walk here between convenience/comprehensiveness in the antora docs, and confining chart-specific details to the chart documentation.  I think what we have is a pretty good split, but I think we could pull it up even a little higher here. For instance instead of getting into "Persistent Volumes" specifically, we can just say the high-level goal:  Attaching the properties to the local file system of the service through a mounted volume.   I don't have great specifics here, but maybe the key question is: Could someone with very little Kubernetes/Helm knowledge read the paragraph and get good understanding of _what's_ happening and why (even if they don't fully understand the "how").

Author: Brought up to higher level by removing kubernetes specific wording.

===== Properties File Locations
Within the configuration store's helm chart, one can specify environment variables with the URIs that specify where the project's various 
properties files are located within the PV. It is required to specify a base property URI that houses the base/default property 
configurations. In addition to the base URI, one can optionally specify a secondary URI for properties files that house the environment
specific configurations. 

NOTE: Multiple configuration properties files can be stored at either URI.

Since it is common practice to define separate Helm chart values for each environment of a project's development lifecycle, it is 
encouraged to define one shared base URI and respective environment specific URIs, each housing the relevant overrides.

The following Helm chart configuration demonstrates the URI specifications for an example CI deployment:
[source,yaml]
----
  env:

>     BASE_PROPERTY: <URI housing base property configurations>

Reviewer: S: This is an example of the type of detail that I think is maybe too much for these docs.  There are two risks with this example: 1) it muddies the waters of the overall explanation of the feature and generally how it works, and 2) it's very likely to fall out of date as the chart changes.

Author: Moved technical portion to helm chart README

    ENVIRONMENT_PROPERTY: <URI housing CI-specific property overrides/augmentations>
----

==== Backend Storage Options
The universal configuration store offers robust production ready configuration storage using https://developer.hashicorp.com/vault[Vault,role=external,window=_blank]. This is an excellent solution for large 

> enterprise applications. For smaller applications looking for a more light weight but still robust storage option, the configuration store 

Reviewer: A: Let's talk about the light-weight solution first and mention it's the default.  And then we can segue into "For enterprise applications that have more robust requirements.."

can be set to use https://github.com/TechnologyBrewery/krausening[Krausening,role=external,window=_blank]. The configuration store can also be set up to use in memory storage, this is best
for quick local testing.

The default property storage is Krausening. To change it to use one of the other options, set the environment
variable `STORAGE_CLASS` to the desired value. The values currently available are `krausening`, `vault`, and `inMemory`.

=== Usage
After the configurations are ingested and reconciled within the store, the next step is to access them within a project using one of 

> the two available methods: the API endpoint or kubernetes resource injection. See the applicable sections below for more details on 

Reviewer: A: Let's make sure we've capitalized proper nouns like _"Kubernetes"_ throughout the page.

utilizing these approaches.

> ==== API Endpoint

Reviewer: A: Maybe "REST Endpoint" instead of API?

The configuration store includes a kubernetes service with the endpoints detailed below. The API endpoint provides a simple yet flexible way 
to access any configurations from within a project.

.Endpoints
[cols="2,1,4"]
|===
| Endpoint Path | HTTP Method | Description

| `/aissemble-properties/{groupName}/{propertyName}`
| `GET`

> | Returns a JSON object containing the value of a property with the given group name and property name.

Reviewer: A: I would define the URL path items as you would Query params here.  Allows us to more cleanly tie the "group name" to the source properties file name.

Author: Updated

| `/aissemble-properties/healthcheck`
| `GET`
| Returns a plain text string used for validating the service is reachable.

|===

Here is an example `curl` request to the configuration store service containing the properties files defined above:
[source,bash]
----
curl http://configuration-store.config-store-ns.svc:8083/aissemble-properties/messaging/topic
----

Resulting JSON object:
[source,json]
----
{
  "groupName": "messaging",
  "name": "topic",
  "value": "ciTopic"
} 
----

> NOTE: The name of the configuration service may vary depending on the projects helm chart configuration, by default it is

Reviewer 2: A: missing apostrophe - project's

Author: Fixed

`http://configuration-store.config-store-ns.svc:8083`

> ==== Resource Injection

Reviewer 2: A: Should this be _Kubernetes_ Resource Injection to match what we called it above?

The configuration store provisions a https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/[kubernetes admission webhook,role=external,window=_blank] as part of its deployment. This webhook is used for injecting property values

> from the configuration store directly into kubernetes resources as they deployed to the cluster.

Reviewer 2: A: recommend minor edit to "...as they _are_ deployed..."

?: Something that struck me as I was reading through this page is we give examples of using what we have built, but don't necessarily give examples of when to use them, or why you would choose one approach over another. You can tell me to pound sand if that is not something we typically do and is why we did not do it here. But I do wonder if providing some help to the user would be useful. E.g., when would they want to use the REST endpoint/API vice Kubernetes Resource Injection; when should they use Krausening vs Vault, and maybe some others. Anyway, it is just a bit of food for thought.

Author: The backend storage section gives a little insight into why you would want to use the different options.

In terms of the REST Endpoint vs kubernetes injection, I think the injection is the preferred method and can be used in replacement of the REST Endpoint in most use cases. However, at the end of the day it really comes down to user preference if they'd rather use the endpoint.

of choice.

==== Defining Configuration Properties
The configuration store utilizes the opensource project https://github.com/TechnologyBrewery/krausening[Krausening,role=external,window=_blank] for managing its configurations, where each configuration is 
represented by a key value pair property. The key values pairs are stored as string types using the standard `.properties` file 
format.

Here is an example properties file:
[source,properties]
----
connector=smallrye-kafka
topic=baseTopic
----
This file defines two configuration properties: `connector` and `topic`.

TIP: For sensitive configuration values, Krausening supports https://github.com/TechnologyBrewery/krausening/blob/dev/krausening/README.md#krausening-in-four-pints-leveraging-jasypt-for-encryptingdecrypting-properties[encrypting and decrypting properties,role=external,window=_blank]

===== Environment Specific Overrides
In addition to the properties file introduced above, one can optionally specify a secondary properties file that houses environment

> specific configurations. These configurations will then used to override their respective base configurations.

Reviewer: A:
```suggestion
specific configurations. These configurations will override their respective base configurations.
```

Continuing with the example above, suppose one defines an additional properties file like the following:
[source,properties]
----
topic=ciTopic
auth-token=auth-secret
----

Then when these two properties files are reconciled within the store, it results in a configuration equivalent to the following:
[source,properties]
----
connector=smallrye-kafka
topic=ciTopic
auth-token=auth-secret
----

NOTE: The `topic` property is overwritten when the environment specific value `ciTopic`.

==== Deploying the Configuration Store

> The https://github.com/boozallen/aissemble/blob/{git-tree}/extensions/extensions-helm/aissemble-configuration-store-chart/README.md[configuration store helm chart,role=external,window=_blank] is used for easily deploying it to a Kubernetes cluster. As part of the deployment, one must

Reviewer: ```suggestion
The https://github.com/boozallen/aissemble/blob/{git-tree}/extensions/extensions-helm/aissemble-configuration-store-chart/README.md[Configuration Store Helm chart,role=external,window=_blank] can easily deploy the store to a Kubernetes cluster. As part of the deployment, one must
```

Reviewer: Not sure if we're capitalizing "Configuration Store". I'm fine with either, but Helm should be capitalized.

Author: Capitalized all helms and config stores

> <<mounting-properties-files,mount the properties files>> to the Kubernetes resource allowing the configuration store to access the custom properties. Additionally,

Reviewer: A:
```suggestion
provide the store access to the custom configurations by <<mounting-properties-files,mount the properties files>> to the Kubernetes resource. Additionally,
```

the <<backend-storage-types,backend storage type>> can be adjusted to fit a project's needs.

==== Mounting Properties Files
Properties files are attached to the local file system of the configuration store within Kubernetes through a mounted volume. To mount
a volume to the configuration store, one must set the helm chart's https://github.com/boozallen/aissemble/blob/{git-tree}/extensions/extensions-helm/aissemble-configuration-store-chart/README.md#persistent-volume-properties[persistent volume properties,role=external,window=_blank]. Then the helm chart's https://github.com/boozallen/aissemble/blob/{git-tree}/extensions/extensions-helm/aissemble-configuration-store-chart/README.md#environment-variables[environment 

> variables,role=external,window=_blank] must be updated with the corresponding URI(s) to the properties files within that volume. It is required to specify 

Reviewer: A: I think it might be better to use the word "location" over "URI", since (at least for now) we are _assuming_ the URI will be a local `file://` URI.

a base property URI that houses the base/default property configurations. In addition to the base URI, one can optionally specify a 
secondary URI for properties files that house the environment specific configurations. 

NOTE: Multiple configuration properties files can be stored at either URI.

=== Backend Storage Types
By default, the configuration store uses Krausening for it backend property storage. This provides a more light weight but still robust 
storage option perfect for smaller applications.

For enterprise applications that have more robust requirements, it is recommended to utilize https://developer.hashicorp.com/vault[Vault,role=external,window=_blank]. Additionally, the configuration 

> store can be set up to use in memory storage, this is best for quick local testing.

Reviewer: ```suggestion
store can be set up to use in-memory storage &mdash; this is best for unit testing.
```

To change it to use one of the other options, set the environment variable `STORAGE_CLASS` to the desired value. The values currently
available are `krausening`, `vault`, and `inMemory`.

== Usage
After the configurations are ingested and reconciled within the store, the next step is to access them within a project using one of 
the two available methods: the <<rest-endpoint,REST endpoint>> or <<kubernetes-resource-injection-recommended,Kubernetes resource injection>>.
Configuration properties are identified by using a group name and property name, where group name is determined by the name of the file
the property is defined in.

Given an example properties file named `messaging.properties`:
[source,properties]
----
connector=smallrye-kafka
----

The property name would be `connector` and group name would be `messaging`.

> === REST Endpoint

Reviewer: Q: Do we have a client API that we want to reference here? (Can't recall if we created one yet.)

Author: At this time we do not
